### PR TITLE
search: exactly one repo disregards default fork or archive exclusion

### PR DESCRIPTION
**Pull request by rvantonder**
Addresses #8953. e2e must pass.

### CODE REVIEW

File: web/src/regression/search.test.ts [hunk 1]
```diff
@@ -432,16 +432,34 @@ describe('Search regression test suite', () => {
...
             await driver.page.goto(config.sourcegraphBaseUrl + '/search?' + urlQuery)
             await driver.page.waitForFunction(() => document.querySelectorAll('.e2e-search-result').length === 0)
         })
-        test('Forked repos included by by fork option', async () => {
+        test('Forked repos included by fork option', async () => {
             const urlQuery = buildSearchURLQuery('type:repo sgtest/mux fork:yes', GQL.SearchPatternType.regexp, false)
             await driver.page.goto(config.sourcegraphBaseUrl + '/search?' + urlQuery)
             await driver.page.waitForFunction(() => document.querySelectorAll('.e2e-search-result').length > 0)
         })
+        test('Single forked repo included if exact', async () => {
```
Comment (unknwon) [inline web/src/regression/search.test.ts]: Hmm, why do we have two "Single forked repo included if exact" cases? From a rookie's eye, the purpose seems identical to me.

Comment (rvantonder) [reply web/src/regression/search.test.ts]: No you just have a good eye. The other is supposed to be archived :) Thanks!

File: web/src/regression/search.test.ts [hunk 2]
```diff
@@ -432,16 +432,34 @@ describe('Search regression test suite', () => {
...
```
Comment (rvantonder) [inline web/src/regression/search.test.ts]:
```suggestion
        test('Single archived repo included if exact', async () => {
```

**Comment by rvantonder** [comment]
I don't understand why [Percy is flaking](https://percy.io/Sourcegraph/Sourcegraph/builds/4503443?utm_campaign=Sourcegraph&utm_content=Sourcegraph&utm_source=github_status_public). Maybe our page takes too long to load the assets.

File: cmd/frontend/graphqlbackend/search.go [hunk 3]
```diff
@@ -256,6 +256,18 @@ func resolveRepoGroups(ctx context.Context) (map[string][]*types.Repo, error) {
...
 }
 
+// exactlyOneRepo returns whether exactly one repo: field is specified and
+// delineated by regex anchors ^ and $. This function helps determine whether we
+// should return results for a single repo regardless of whether it is a fork or
+// archive.
+func exactlyOneRepo(repoFilters []string) bool {
+	if len(repoFilters) == 1 {
+		filter := repoFilters[0]
+		return strings.HasPrefix(filter, "^") && strings.HasSuffix(filter, "$")
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search.go]: should we be parsing `filter` and checking that it is a literal with anchors? IE this check passes `^.*$`. This is likely overkill and its fine for our magic to false fire in those unlikely circumstances.

File: cmd/frontend/graphqlbackend/search.go [hunk 4]
```diff
@@ -285,14 +297,14 @@ func (r *searchResolver) resolveRepositories(ctx context.Context, effectiveRepoF
 
 	forkStr, _ := r.query.StringValue(query.FieldFork)
 	fork := parseYesNoOnly(forkStr)
-	if fork == Invalid {
-		fork = No // fork defaults to No.
+	if fork == Invalid && !exactlyOneRepo(repoFilters) {
+		fork = No // fork defaults to No unless exactly one repo is being searched.
```
Comment (keegancsmith) [inline cmd/frontend/graphqlbackend/search.go]: previously `fork == Invalid` never survived,but now we can have a fork value of `Invalid`. What does `fork == Invalid` mean when passed to our repo search layer?

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go]: Previous to changing the fork default behavior in https://github.com/sourcegraph/sourcegraph/pull/8739, `Invalid` did survive, this was the result of `parseYesNoOnly` when `fork` was not specified. So it does whatever happened prior to these changes (i.e., forks included). "Whatever" specifically means we evaluate [these fields](https://sourcegraph.com/github.com/sourcegraph/sourcegraph@9d65befdb15a677285b2826cb5e59a9d27f8ec8b/-/blob/cmd/frontend/graphqlbackend/search.go#L317-318) in `resolveRepositories` as we would have when forks were included.

File: cmd/frontend/graphqlbackend/search.go (hunk 3 above)
Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go]: Yeah, makes sense. See upcoming PR for an idea on how to check this; I don't think there's a nice existing utility function.

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go]: Yeah, the main problem is that something like `github\.com/sourcegraph/zoekt` for our purposes is a "literal" string but this _isn't_ a literal string, it contains regex escape characters. My idea was to check that `filter == regexp.QuoteMeta(filter)` but `QuoteMeta` will escape the `\`.

Comment (rvantonder) [reply cmd/frontend/graphqlbackend/search.go]: Nvm I discovered `syntax.OpLiteral` in regexp/syntax.

**Pull request merged**